Add flagging and correct deltaFwd/deltaBwd computation for tracks crossing CE

davidrohr opened the pull request:
@shahor02 : This PR implements the flagging on the O2 side. It needs AliTPCCommon v.1.4.4 (alisw/alidist#1315).
Marked as [WIP], as the delta computation is not tested yet.

davidrohr commented:
fails as alisw/alidist#1315 is not yet merged.

shahor02 commented on Detectors/TPC/reconstruction/src/TPCCATracking.cxx:
```diff
@@ -286,6 +304,10 @@ int TPCCATracking::runTracking(const ClusterNativeAccessFullTPC& clusters, std::
+        oTrack.setHasCSideClusters();
+        oTrack.setHasASideClusters();
+      }
       if (tracks[i].CSide()) {
```
should not it be ?
```
else {
  if (tracks[i].CSide()) {
     oTrack.setHasCSideClusters();
   } else {
      oTrack.setHasASideClusters();
   }
}
```

davidrohr replied:
does not make a difference in reality, but anyway, your version is cleaner so I changed it.

shahor02 commented on Detectors/TPC/reconstruction/src/TPCCATracking.cxx:
```diff
@@ -248,28 +248,46 @@ int TPCCATracking::runTracking(const ClusterNativeAccessFullTPC& clusters, std::
         float zoffset = tracks[i].CSide() ? -tracks[i].GetParam().GetZOffset() : tracks[i].GetParam().GetZOffset();
         time0 = sContinuousTFReferenceLength - zoffset * vzbinInv;
```
This time0 is then set to the track and used in the matching to define the time bracket for the track as
```
    float time0 = trcOrig.getTime0() - mNTPCBinsFullDrift;
    trc.timeBins.set(time0 - trcOrig.getDeltaTBwd() - mTPCTimeEdgeTSafeMargin,
                     time0 + trcOrig.getDeltaTFwd() + mTPCTimeEdgeTSafeMargin);
```
For the CCE tracks time0 calculated with request of the track converging to the vertex will be certainly wrong.
Should not the time0 in this case be calculated as  sContinuousTFReferenceLength + average time of CE-closest clusters?

davidrohr replied:
In principle, the tracker should already shift the track to exactly this offset, so this should not be needed. I'd rather have the correct shift already in the tracking.
However, I'd like to test this in O2. How can I simulate events with tracks crossing the CE?

shahor02 replied:
I mean the CE-crossing tracks cannot converge to Z=0 in principle (except for the strictly eta=0 tracks), so their shifting to Z=0 will move the clusters close to the CE to wrong side.
You should get such tracks in any simulation with non-0 vertex spread and smooth eta distribution, i.e. `o2sim -n50 -g pythia8` should produce them.

davidrohr commented:
This has been tested, and could be merged principally.
However, I am waiting for the next tag of AliTPCCommon to be published.
The track merging and flagging works correctly across the CE, but due a problem in digitization/clusterization there are too many split clusters and the tracks are bad.
This needs to be fixed before the new features can be used realistically.

matthiasrichter commented on Detectors/TPC/reconstruction/src/TPCCATracking.cxx:
```diff
@@ -248,28 +248,46 @@ int TPCCATracking::runTracking(const ClusterNativeAccessFullTPC& clusters, std::
+          for (int iCl = 1;iCl < tracks[i].NClusters();iCl++)
+          {
+            if (lastSide ^ (trackClusters[tracks[i].FirstClusterRef() + iCl].fSlice < 18))
+            {
```
same here (bracket and `Sector::MAXSECTOR / 2`)

matthiasrichter commented on Detectors/TPC/reconstruction/src/TPCCATracking.cxx:
```diff
@@ -248,28 +248,46 @@ int TPCCATracking::runTracking(const ClusterNativeAccessFullTPC& clusters, std::
-          tFwd = (detParam.getTPClength() - zHighAbs) * vzbinInv;
-          tBwd = zLowAbs * vzbinInv;
+        if (tracks[i].CCE()) {
+          bool lastSide = trackClusters[tracks[i].FirstClusterRef()].fSlice < 18;
```
replace 18 with `Sector::MAXSECTOR / 2` as further down

davidrohr commented:
@matthiasrichter : pushed new version that addresses your comments and fully fixes the new features.

This shall be merged only after alisw/alidist#1405 and #1497 as it relies on changes in AliTPCCommon.

davidrohr commented:
@matthiasrichter : The dependent PRs have been merged, so from my side this can be merged
